### PR TITLE
🎉Source Google Analytics Data Api: code refactoring

### LINKED ISSUE #6326: Source Google Analytics v4: allow run GA account without view id
## Tell us about the problem you're trying to solve
From slack convo:

Is this your first time deploying Airbyte: Yes
OS Version / Instance: Ubuntu 20.04
Memory / Disk: 8Gb / 160GB SSD
Deployment: Docker
Airbyte Version: 0.29.2-alpha
Source name/version: Google Analytics v4 0.1.1
Destination name/version: Postgres 0.3.10
Step: Create New Source
Description: Source connector require View ID but GA4 doesn't have View. Since this connector is v4 so how can I pull data from my website that uses GA4 instead of UA.
This problem happens all version. Our Google Analytics connector and Google Analytics v4 connector both require View ID. But View ID only exist on Google Analytics properties using UA tag. Google Analytics v4 now offers properties with GA4 tag. Properties with GA4 tag don't have View so I cannot use our connectors to pull data. I think we need to update our connector.
I don't know how to create a Github issue though smh.

## Describe the solution you’d like
image
I don't know if only removing the VIEW_ID field as required will work.

## Describe the alternative you’ve considered or used
A clear and concise description of any alternative solutions or features you've considered or are using today.

## Are you willing to submit a PR?

Remove this with your answer :-)

### PR DESCRIPTION
**Pull request by roman-yermilov-gl**
## What
*Refactoring of the Google analytics data API source. Use REST API calls instead of library and support a default set of reports*

## Pre-merge Checklist
Expand the relevant checklist and delete the others.

<details><summary><strong>Updating a connector</strong></summary>

### Community member or Airbyter

- [ ] Grant edit access to maintainers ([instructions](https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests))
- [ ] Secrets in the connector's spec are annotated with `airbyte_secret`
- [ ] Unit & integration tests added and passing. Community members, please provide proof of success locally e.g: screenshot or copy-paste unit, integration, and acceptance test output. To run acceptance tests for a Python connector, follow instructions in the README. For java connectors run `./gradlew :airbyte-integrations:connectors:<name>:integrationTest`.
- [ ] Code reviews completed
- [ ] Documentation updated
    - [ ] Connector's `README.md`
    - [ ] Connector's `bootstrap.md`. See [description and examples](https://docs.google.com/document/d/1ypdgmwmEHWv-TrO4_YOQ7pAJGVrMp5BOkEVh831N260/edit?usp=sharing)
    - [ ] Changelog updated in `docs/integrations/<source or destination>/<name>.md` including changelog. See changelog [example](https://docs.airbyte.io/integrations/sources/stripe#changelog)
- [ ] PR name follows [PR naming conventions](https://docs.airbyte.io/contributing-to-airbyte/updating-documentation#issues-and-pull-requests)

### Airbyter

If this is a community PR, the Airbyte engineer reviewing this PR is responsible for the below items.

- [ ] Create a non-forked branch based on this PR and test the below items on it
- [ ] Build is successful
- [ ] If new credentials are required for use in CI, add them to GSM. [Instructions](https://docs.airbyte.io/connector-development#using-credentials-in-ci).
- [ ] [`/test connector=connectors/<name>` command](https://docs.airbyte.io/connector-development#updating-an-existing-connector) is passing
- [ ] New Connector version released on Dockerhub and connector version bumped by running the `/publish` command described [here](https://docs.airbyte.io/connector-development#updating-an-existing-connector)

</details>

## Tests

<details><summary><strong>Unit</strong></summary>

*Put your unit tests output here.*

</details>

<details><summary><strong>Integration</strong></summary>

*Put your integration tests output here.*

</details>

<details><summary><strong>Acceptance</strong></summary>

*Put your acceptance tests output here.*

</details>

### CODE REVIEW

File: airbyte-webapp/src/hooks/services/Feature/constants.ts [hunk 1]
```diff
@@ -7,4 +7,5 @@ export const defaultFeatures = [
   FeatureItem.AllowSync,
   FeatureItem.AllowUpdateConnectors,
   FeatureItem.AllowUploadCustomImage,
+  FeatureItem.AllowOAuthConnector,
```
Comment (edmundito) [inline airbyte-webapp/src/hooks/services/Feature/constants.ts:10]: This is not a default feature and should be removed.

Comment (lazebnyi) [reply airbyte-webapp/src/hooks/services/Feature/constants.ts]: updated.

File: airbyte-integrations/connectors/source-google-analytics-data-api/Makefile [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+	@docker run --rm -v $(PWD)/secrets:/secrets $(docker_image) check --config /secrets/config.json | jq
+
+discover:
+	@docker run --rm -v $(PWD)/secrets:/secrets $(docker_image) discover --config /secrets/config.json | jq
+
+read:
+	@docker run --rm -v $(PWD)/secrets:/secrets -v $(PWD)/integration_tests:/integration_tests $(docker_image) read --config /secrets/config.json --catalog /integration_tests/configured_catalog.json | jq
+
+unittest-local:
+	@python -m pytest unit_tests
```
Comment (Phlair) [inline airbyte-integrations/connectors/source-google-analytics-data-api/Makefile:19]: what's this Makefile for?

File: airbyte-integrations/connectors/source-google-analytics-data-api/setup.py [hunk 3]
```diff
@@ -5,7 +5,7 @@
 
 from setuptools import find_packages, setup
 
-MAIN_REQUIREMENTS = ["airbyte-cdk~=0.1", "google-analytics-data==0.11.2"]
+MAIN_REQUIREMENTS = ["airbyte-cdk~=0.1", "google-analytics-data==0.11.2", "PyJWT", "cryptography", "requests"]
```
Comment (Phlair) [inline airbyte-integrations/connectors/source-google-analytics-data-api/setup.py:8]: lib versions should be pinned please

File: airbyte-integrations/connectors/source-google-analytics-data-api/source_google_analytics_data_api/authenticator.py [hunk 4]
```diff
@@ -0,0 +1,75 @@
+import datetime
+
+import jwt
+import requests
+
+from source_google_analytics_data_api import utils
+
+
+class GoogleServiceKeyAuthenticator(requests.auth.AuthBase):
```
Comment (Phlair) [inline airbyte-integrations/connectors/source-google-analytics-data-api/source_google_analytics_data_api/authenticator.py:9]: is this completely unique such that we can't use any of the cdk authenticator frameworks?

File: airbyte-integrations/connectors/source-google-analytics-data-api/source_google_analytics_data_api/utils.py [hunk 5]
```diff
@@ -0,0 +1,14 @@
...
+def datetime_to_secs(dt: datetime.datetime) -> int:
+    return calendar.timegm(dt.utctimetuple())
+
+
+def string_to_date(d: str, f: str = "%Y-%m-%d") -> datetime.date:
+    return datetime.datetime.strptime(d, f).date()
+
+
+def date_to_string(d: datetime.date, f: str = "%Y-%m-%d") -> str:
+    return d.strftime(f)
```
Comment (Phlair) [inline airbyte-integrations/connectors/source-google-analytics-data-api/source_google_analytics_data_api/utils.py:14]: nit: we generally use `pendulum` library for datetime operations (it's basically datetime++)

**Review by Phlair — COMMENTED** [review]
Tests + Oauth implementation looking great!

Some open questions...

File: airbyte-integrations/connectors/source-google-analytics-data-api/Makefile (hunk 2 above)
Comment (roman-yermilov-gl) [reply airbyte-integrations/connectors/source-google-analytics-data-api/Makefile:19]: It is not necessary to have it in repository. I created it for running routine operations more quickly. It is only for development purpose and for convenience. For example `check` is a long command but I am just typing `make check` in my console

File: airbyte-integrations/connectors/source-google-analytics-data-api/source_google_analytics_data_api/authenticator.py (hunk 4 above)
Comment (roman-yermilov-gl) [reply airbyte-integrations/connectors/source-google-analytics-data-api/source_google_analytics_data_api/authenticator.py:9]: I didn't find any proper base class in `cdk` for service authentication. In some other connectors they are using OAuth class as a base class and re-implementing methods but it is basically the same as I do using `requests.auth.AuthBase`. It is more clea[n/r] I think

File: airbyte-integrations/connectors/source-google-analytics-data-api/source_google_analytics_data_api/utils.py (hunk 5 above)
Comment (roman-yermilov-gl) [reply airbyte-integrations/connectors/source-google-analytics-data-api/source_google_analytics_data_api/utils.py:14]: Thanks, I will keep it in mind. Should I rewrite these utils for using `pendulum` or they are good as is for just doing formatting?

File: airbyte-integrations/connectors/source-google-analytics-data-api/setup.py (hunk 3 above)
Comment (roman-yermilov-gl) [reply airbyte-integrations/connectors/source-google-analytics-data-api/setup.py]: done

File: airbyte-integrations/connectors/source-google-analytics-data-api/source_google_analytics_data_api/utils.py (hunk 5 above)
Comment (Phlair) [reply airbyte-integrations/connectors/source-google-analytics-data-api/source_google_analytics_data_api/utils.py:14]: this is fine if it works

**Review by Phlair — APPROVED** [review]
Nice, lgtm 👍

**Review by edmundito — COMMENTED** [review]
Thanks for removing the frontend change ✅

**Review by arsenlosenko — APPROVED** [review]

**Comment by grubberr** [comment]
@roman-yermilov-gl 
I think we don't need this
`airbyte-integrations/connectors/source-google-analytics-data-api/integration_tests/catalog.json`

**Review by bazarnov — APPROVED** [review]

**Review by grubberr — APPROVED** [review]

**Comment by roman-yermilov-gl** [comment]
/publish connector=connectors/source-google-analytics-data-api

**Pull request merged**